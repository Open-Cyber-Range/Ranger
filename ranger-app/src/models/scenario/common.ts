export type Source = {
  name: string;
  version: string;
};
