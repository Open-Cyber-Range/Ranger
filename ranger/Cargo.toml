[package]
name = "ranger"
version = "2.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[package.metadata.deb]
name = "ranger-server"
maintainer = "C14 developers <developers@cr14.ee>"
copyright = "MIT"
depends = "$auto, systemd"
section = "utility"
priority = "optional"
changelog = "changelog"
assets = [
  [
    "target/release/ranger",
    "/var/opt/ranger/bin/",
    "755",
  ],
  [
    "assets/default-production-config.yml",
    "/etc/opt/ranger/ranger/config.yml",
    "644",
  ],
  [
    "../ranger-app/build/assets/*",
    "/etc/opt/ranger/ranger-app/assets",
    "644",
  ],
  [
    "../ranger-app/build/*",
    "/etc/opt/ranger/ranger-app/",
    "644",
  ],
]
systemd-units = { unit-name = "ranger", unit-scripts = "assets/unit-scripts" }

[dependencies]
actix = "0.13"
actix-multipart = "0.6"
actix-web = "4"
actix-web-actors = "4"
actix-web-grants = "3"
actix-http = { version = "3", features = ["http2", "ws"] }
actix-web-httpauth = "0.8"
anyhow = "1"
async-trait = "0.1"
bigdecimal = { version = "0.4", features = ["serde"] }
chrono = { version = "0.4", features = ["serde"] }
diesel = { version = "2.1", features = ["mysql", "r2d2", "chrono", "numeric"] }
diesel_migrations = "2.1"
env_logger = "0.9"
futures = "0.3"
futures-util = "0.3"
jsonwebtoken = "8"
keycloak = { git = "https://github.com/ebakoba/keycloak", branch = "master" }
log = "0.4"
lazy_static = "1.4"
lettre = { version = "0.10", default-features = false, features = [
  "smtp-transport",
  "pool",
  "hostname",
  "builder",
  "rustls",
  "rustls-tls",
] }
sha3 = "0.10"
digest = "0.10"
rand = "0.8"
ranger-grpc = { path = "../ranger-grpc" }
reqwest = { version = "0.11", default-features = false, features = [
  "json",
  "rustls",
  "rustls-tls",
  "tokio-rustls",
  "hyper-rustls",
  "rustls-native-certs",
] }
sdl-parser = "0"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1"
serde_yaml = "0.9"
semver = "1"
tempfile = "3"
thiserror = "1"
toml-query = "0.10"
toml = "0.5.6"
tokio = { version = "1", features = [
  "rt-multi-thread",
  "time",
  "fs",
  "macros",
  "net",
  "rt",
] }
tonic = "0.8"
uuid = { version = "1.1", features = ["v4", "serde", "fast-rng"] }

[dev-dependencies]
actix-rt = "2"
insta = { version = "1", features = ["redactions", "yaml"] }
sdl-parser = { version = "0", features = ["test"] }
