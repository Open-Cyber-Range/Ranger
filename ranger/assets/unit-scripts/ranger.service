[Unit]
Description=ranger
After=network.target

[Service]
ExecStart=/var/opt/ranger/bin/ranger /etc/opt/ranger/ranger/config.yml
Type=simple
Restart=on-failure
RestartSec=10
StartLimitInterval=10m
StartLimitBurst=5

[Install]
WantedBy=multi-user.target
