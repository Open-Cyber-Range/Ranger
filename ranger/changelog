ranger (0.2.0) stretch; urgency=medium

  * Initial Release
  * Adds web application support

 -- Kaarel Allemann <kaarel.allemann@cr14.ee>  Sat, 05 Nov 2022 19:23:51 +0200
