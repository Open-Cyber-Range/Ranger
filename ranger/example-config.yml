host: localhost
port: 8080
repository:
  folder: /tmp/test-repo
  username: some-username
  email: some@email.com
package_folder: /tmp/packages
