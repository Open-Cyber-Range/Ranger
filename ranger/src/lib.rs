pub(crate) mod configuration;
mod constants;
pub mod errors;
pub mod models;
pub mod routes;
pub(crate) mod schema;
pub(crate) mod services;
pub(crate) mod utilities;

use crate::services::database::Database;
use actix::{Actor, Addr};
use anyhow::Result;
use configuration::{read_configuration, Configuration};
use log::{error, info};
use services::{
    deployer::{DeployerDistribution, DeployerFactory},
    deployment::DeploymentManager,
    mailer::Mailer,
    scheduler::Scheduler,
    websocket::WebSocketManager,
};

pub struct AppState {
    pub database_address: Addr<Database>,
    pub deployment_manager_address: Addr<DeploymentManager>,
    pub configuration: Configuration,
    pub websocket_manager_address: Addr<WebSocketManager>,
}

impl AppState {
    pub fn new(
        configuration: &configuration::Configuration,
        distributor: &Addr<DeployerDistribution>,
        database: &Addr<Database>,
        websocket_manager: &Addr<WebSocketManager>,
    ) -> Self {
        let schduler_address = Scheduler::new().start();
        AppState {
            database_address: database.clone(),
            deployment_manager_address: DeploymentManager::new(
                schduler_address,
                distributor.clone(),
                database.clone(),
                configuration.deployment_groups.clone(),
                configuration.default_deployment_group.to_string(),
            )
            .start(),
            configuration: configuration.to_owned(),
            websocket_manager_address: websocket_manager.clone(),
        }
    }
}

pub async fn app_setup(environment_arguments: Vec<String>) -> Result<(String, u16, AppState)> {
    let configuration = read_configuration(environment_arguments)?;
    let deployer_factory = DeployerFactory::new(&configuration.deployers)
        .await
        .unwrap_or_else(|error| panic!("Failed to create deployer distribution: {error}"))
        .start();

    let deployer_distributor = DeployerDistribution::new(
        deployer_factory,
        configuration.deployers.keys().cloned().collect(),
    )
    .await
    .unwrap_or_else(|error| panic!("Failed to create deployer distribution: {error}"))
    .start();

    let websocket_manager = WebSocketManager::new().start();
    let database = Database::try_new(&configuration.database_url, &websocket_manager)
        .unwrap_or_else(|error| {
            panic!(
                "Failed to create database connection to {} due to: {error}",
                &configuration.database_url
            )
        })
        .start();

    send_mail(configuration.clone());

    let app_state = AppState::new(
        &configuration,
        &deployer_distributor,
        &database,
        &websocket_manager,
    );
    Ok((configuration.host, configuration.port, app_state))
}

pub fn send_mail(configuration: Configuration) {
    if let Some(mailer_configuration) = configuration.mailer_configuration {
        match Mailer::send_mail(
            mailer_configuration,
            "real.person@real-email.com".to_string(),
        ) {
            Ok(_) => info!("Mail sent successfully!"),
            Err(e) => error!("Mailer failed: {:?}", e),
        }
    }
}
