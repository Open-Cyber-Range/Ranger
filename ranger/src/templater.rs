use std::collections::HashMap;

use actix::{Actor, Addr, Context, Handler, Message, ResponseFuture};
use anyhow::{anyhow, Ok, Result};
use async_trait::async_trait;
use futures::future::join_all;
use futures::Future;
use log::error;
use ranger_grpc::NodeDeployment;
use ranger_grpc::{
    template_service_client::TemplateServiceClient, Identifier, Source as GrpcSource,
};
use sdl_parser::node::Source;
use sdl_parser::Scenario;
use tonic::transport::Channel;

use crate::machiner::NodeDeploymentTrait;

#[derive(Message)]
#[rtype(result = "Result<Identifier, anyhow::Error>")]
pub struct CreateTemplate(pub GrpcSource);

#[derive(Message)]
#[rtype(result = "Result<(), anyhow::Error>")]
pub struct DeleteTemplate(pub Identifier);

pub struct TemplateClient {
    template_client: TemplateServiceClient<Channel>,
}

#[async_trait]
trait DeployerActions<T>
where
    Self: Actor,
{
    async fn create_template(templater_address: &Addr<Self>, create_info: T) -> Result<String>;
}
#[async_trait]
impl DeployerActions<Source> for TemplateClient {
    async fn create_template(
        templater_address: &Addr<TemplateClient>,
        source: Source,
    ) -> Result<String> {
        Ok(templater_address
            .send(CreateTemplate(GrpcSource {
                name: source.name.clone(),
                version: source.version.clone(),
            }))
            .await??
            .value)
    }
}

pub fn initiate_template_clients(
    deployers: HashMap<String, String>,
) -> Vec<impl Future<Output = Result<(String, Addr<TemplateClient>), anyhow::Error>>> {
    deployers
        .into_iter()
        .map(|(name, ip)| async {
            Ok::<(String, Addr<TemplateClient>)>((name, TemplateClient::new(ip).await?.start()))
        })
        .collect()
}

pub type TemplateClientResults = Vec<Result<(String, Addr<TemplateClient>), anyhow::Error>>;
pub trait TemplateClientFilter {
    fn filter_template_clients(self) -> HashMap<String, Addr<TemplateClient>>;
}
impl TemplateClientFilter for TemplateClientResults {
    fn filter_template_clients(self) -> HashMap<String, Addr<TemplateClient>> {
        self.into_iter()
            .filter_map(|node_client| {
                node_client
                    .map_err(|error| error!("Error setting up TemplateClient: {}", error))
                    .ok()
            })
            .collect::<HashMap<String, Addr<TemplateClient>>>()
    }
}

pub fn separate_node_deployments_by_type(
    node_deployments: Vec<NodeDeployment>,
) -> Result<(Vec<NodeDeployment>, Vec<NodeDeployment>)> {
    let mut machiner_deployments = vec![];
    let mut switcher_deployments = vec![];
    for node_deployment in node_deployments {
        match node_deployment
            .clone()
            .node
            .ok_or_else(|| anyhow!("Error getting node"))?
            .identifier
            .ok_or_else(|| anyhow!("Error getting identifier"))?
            .node_type()
        {
            ranger_grpc::NodeType::Vm => machiner_deployments.push(node_deployment),
            ranger_grpc::NodeType::Switch => switcher_deployments.push(node_deployment),
        }
    }
    Ok((machiner_deployments, switcher_deployments))
}
pub fn filter_node_deployments(
    node_deployment_results: Vec<Result<NodeDeployment, anyhow::Error>>,
) -> Result<Vec<NodeDeployment>> {
    let node_deployments: Vec<NodeDeployment> = node_deployment_results
        .into_iter()
        .filter_map(|node_deployment| {
            node_deployment
                .map_err(|error| error!("Error creating node deployment: {}", error))
                .ok()
        })
        .collect();
    if node_deployments.is_empty() {
        Err(anyhow!("No nodes to deploy"))
    } else {
        Ok(node_deployments)
    }
}

pub async fn create_node_deployments(
    scenario: Scenario,
    templaters: &HashMap<String, Addr<TemplateClient>>,
    exercise_name: &str,
) -> Result<Vec<Result<NodeDeployment>>> {
    let nodes = scenario.nodes.ok_or_else(|| anyhow!("Nodes not found"))?;
    let node_deployments = join_all(nodes.iter().zip(templaters.iter().cycle()).map(
        |((node_name, node), (_, templater_address))| async move {
            let source = node
                .source
                .as_ref()
                .ok_or_else(|| anyhow!("Source not found"))?;
            let template_id =
                TemplateClient::create_template(templater_address, source.clone()).await?;
            let node_deployment = match node.type_field {
                sdl_parser::node::NodeType::VM => NodeDeployment::default().initialize_vm(
                    node.clone(),
                    node_name.to_string(),
                    template_id,
                    exercise_name.to_string(),
                )?,
                sdl_parser::node::NodeType::Network => NodeDeployment::default()
                    .initialize_switch(
                        node_name.to_string(),
                        template_id,
                        exercise_name.to_string(),
                    )?,
            };
            Ok::<NodeDeployment>(node_deployment)
        },
    ))
    .await;
    Ok(node_deployments)
}

impl TemplateClient {
    pub async fn new(server_address: String) -> Result<Self> {
        Ok(Self {
            template_client: TemplateServiceClient::connect(server_address.clone()).await?,
        })
    }
}

impl Actor for TemplateClient {
    type Context = Context<Self>;
}

impl Handler<CreateTemplate> for TemplateClient {
    type Result = ResponseFuture<Result<Identifier>>;

    fn handle(&mut self, msg: CreateTemplate, _ctx: &mut Self::Context) -> Self::Result {
        let template_deployment = msg.0;
        let mut client = self.template_client.clone();
        Box::pin(async move {
            let result = client
                .create(tonic::Request::new(template_deployment))
                .await?;
            Ok(result.into_inner())
        })
    }
}

impl Handler<DeleteTemplate> for TemplateClient {
    type Result = ResponseFuture<Result<()>>;

    fn handle(&mut self, msg: DeleteTemplate, _ctx: &mut Self::Context) -> Self::Result {
        let identifier = msg.0;
        let mut client = self.template_client.clone();
        Box::pin(async move {
            let result = client.delete(tonic::Request::new(identifier)).await;
            if let Err(status) = result {
                return Err(anyhow::anyhow!("{:?}", status));
            }
            Ok(())
        })
    }
}
