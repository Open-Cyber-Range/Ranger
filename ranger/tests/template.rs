mod common;

#[cfg(test)]
mod tests {
    use crate::common::create_mock_vmware_server;
    use actix::Actor;
    use actix_rt::System;
    use anyhow::{Error, Result};
    use ranger::templater::{CreateTemplate, DeleteTemplate, TemplateClient};
    use ranger_grpc::{Identifier, Source};

    #[test]
    fn template_created_successfully() -> Result<()> {
        let socket_address = create_mock_vmware_server().run_template_server()?;
        let system = System::new();
        let template_id = system.block_on(async {
            let template_deployer_client =
                TemplateClient::new(format!("http://{}", socket_address).clone())
                    .await?
                    .start();
            let template_id = template_deployer_client
                .send(CreateTemplate(Source {
                    name: String::from("some-name"),
                    version: String::from("0.1.0"),
                }))
                .await??;
            Ok::<Identifier, Error>(template_id)
        })?;

        insta::assert_debug_snapshot!(template_id);
        Ok(())
    }

    #[test]
    fn template_creation_failed() -> Result<()> {
        let socket_address = create_mock_vmware_server()
            .successful_create(false)
            .run_template_server()?;
        let system = System::new();
        let result = system.block_on(async {
            let template_deployer_client =
                TemplateClient::new(format!("http://{}", socket_address).clone())
                    .await?
                    .start();
            template_deployer_client
                .send(CreateTemplate(Source {
                    name: String::from("some-name"),
                    version: String::from("0.1.0"),
                }))
                .await??;
            Ok::<(), Error>(())
        });

        assert!(result.is_err());
        Ok(())
    }

    #[test]
    fn template_deleted_successfully() -> Result<()> {
        let socket_address = crate::common::create_mock_vmware_server().run_template_server()?;
        let system = System::new();
        system.block_on(async {
            let template_deployer_client =
                TemplateClient::new(format!("http://{}", socket_address).clone())
                    .await?
                    .start();
            template_deployer_client
                .send(DeleteTemplate(Identifier {
                    value: String::from("some-identifier"),
                }))
                .await??;
            Ok::<(), Error>(())
        })?;

        Ok(())
    }

    #[test]
    fn template_deletion_failed() -> Result<()> {
        let socket_address = create_mock_vmware_server()
            .successful_delete(false)
            .run_template_server()?;
        let system = System::new();
        let result = system.block_on(async {
            let template_deployer_client =
                TemplateClient::new(format!("http://{}", socket_address).clone())
                    .await?
                    .start();
            template_deployer_client
                .send(DeleteTemplate(Identifier {
                    value: String::from("some-identifier"),
                }))
                .await??;
            Ok::<(), Error>(())
        });
        assert!(result.is_err());
        Ok(())
    }
}
