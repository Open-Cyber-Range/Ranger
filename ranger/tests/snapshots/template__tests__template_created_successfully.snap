---
source: ranger/tests/template.rs
expression: template_id
---
Identifier {
    value: "Some UUID",
}
